### PR TITLE
Add a checkbox to toggle Tool hotkeys.

### PR DESCRIPTION
**Pull request by AndyZe**
A followup to #1238

### CODE REVIEW

**Comment by AndyZe** [comment]
Would love to get a review before the Melodic release. Should be quick...

File: src/rviz/tool.h [hunk 1]
```diff
@@ -175,7 +177,6 @@ Q_OBJECT
   DisplayContext* context_;
 
   char shortcut_key_;
-  bool access_all_keys_;
```
Comment (dhood) [inline src/rviz/tool.h]: please create getters/setters for this instead of making it public

File: src/rviz/visualization_frame.h [hunk 2]
```diff
@@ -32,6 +32,7 @@
 
 #include <boost/shared_ptr.hpp>
 
+#include <QCheckBox>
```
Comment (dhood) [inline src/rviz/visualization_frame.h]: is this change just leftover from https://github.com/ros-visualization/rviz/pull/1238?

File: src/rviz/visualization_frame.cpp [hunk 3]
```diff
@@ -598,6 +597,11 @@ void VisualizationFrame::openPreferencesDialog()
   if( dialog->exec() == QDialog::Accepted ) {
     // Apply preferences.
     preferences_ = boost::make_shared<Preferences>( temp_preferences );
+
+    if (preferences_->allow_tool_kb_shortcuts)
```
Comment (dhood) [inline src/rviz/visualization_frame.cpp]: please add curly braces around all the if/else statements

**Comment by AndyZe** [comment]
FYI I won't be making any more PR's until this gets approved. I think that's fair.

The hotkeys are very annoying to me and I think there are hundreds of users out there who don't realize why their interactive markers have suddenly disappeared. (Hint: it's because they pressed a hot key and didn't realize it)

**Comment by wjwwood** [comment]
> FYI I won't be making any more PR's until this gets approved. I think that's fair.

We'll get to this pull request as soon as we have some time to do so. I'm sorry to hear you won't be making any additional pull requests (assuming you have some ready) until this is approved, but if your intention is to make this go faster or get us to pay more attention to it by withholding future improvements, that's not going to work because it doesn't give us any extra time in which to look at these things.

> The hotkeys are very annoying to me and I think there are hundreds of users out there who don't realize why their interactive markers have suddenly disappeared. (Hint: it's because they pressed a hot key and didn't realize it)

Our inability to get this in before the release or having time since to look at it has nothing to do with how useful it is, and everything to do with time constraints. We will get to it when we have time, that's the best we can do.

Sorry, and thanks for your patience.

**Comment by AndyZe** [comment]
Yep. I'm not trying to be a dick, but there's no point in making PRs that get reviewed but not merged. Or not even reviewed.

**Comment by AndyZe** [comment]
Maybe just pass the concept that the natives are restless up the command chain...

**Comment by awatson3** [comment]
Status?

**Comment by AndyZe** [comment]
They are too busy even to hire an intern to review these minor pull requests.

**Comment by rhaschke** [comment]
Dear @AndyZe,

I took over maintainership for rviz recently and I'm trying to merge all open PRs.
As this one had conflicts, I rebased and cleaned up a little bit in my own fork: https://github.com/rhaschke/rviz/tree/toggle-hotkeys

Doing so, I stumbled over your commit message in https://github.com/ros-visualization/rviz/pull/1243/commits/78b7c2f45610b894dcaec25a795f77c7198bcfcd:

> Disable hotkeys works with the Preferences dialog, **but logic is flipped**?

Looking into this issue in more detail, I think that `Tool::access_all_keys_` is the wrong variable to use.
This variable [determines whether key presses should be handled by the current tool](https://github.com/ros-visualization/rviz/blob/3a41f37bdfe0959f305ab9a50e7aebdfcf1642db/src/rviz/tool_manager.cpp#L156-L161), even though the pressed key corresponds to a selection hotkey for another tool. This is a private property of the specific tool and, in my opinion, completely unrelated to your goal of disabling all tool selection hotkeys.
I'm not even sure that this was your actual goal. _Selecting_ the `NavGoal` tool and actually _sending_ a goal message should be two different things, right?

A really clean solution would be to make hotkeys completely configurable. Then, one could selectively disable specific hotkeys only.

**Comment by AndyZe** [comment]
:+1:  I'm glad somebody responsive took ownership!  Unfortunately my personal motivation and available time to fix this issue is gone  :/

**Comment by rhaschke** [comment]
> Unfortunately my personal motivation and available time to fix this issue is gone.

While I can understand your disappointment, please contribute to this issue.
I "just" took maintainership to save this project from death, although I'm already heavily busy with maintaining MoveIt! and several other projects. I cannot to this job without the active help of the community.
I thought that I could quickly merge this PR after resolving the conflicts. But I think, this PR is conceptually wrong. Either you can clarify, why I am wrong or you can propose a valid solution.

**Comment by AndyZe** [comment]
Like many people, I often type in a different window while I have RViz open, and that was causing some safety issues. For my use case, I wanted a way to disable ALL hotkeys.

I agree that enabling or disabling individual hotkeys would be a nice feature and it would also accomplish what I needed. Is it worth implementing at this stage in RViz for ROS1? How would it be implemented? I'm not sure

**Comment by rhaschke** [comment]
You are probably right that nobody has resources to implement individual hotkeys...
However, your current approach doesn't work out as well. I was thinking about a global flag used in ToolManager to suppress all hotkeys.

Could you please explain in more detail your safety issue. Just pressing 'g' only switches to the NavGoal tool. To actually send a goal, you also need to interact with the mouse, don't you?

By the way, another option to disable NavGoal, is to simply remove the tool. Would this be an option for you as well?

**Comment by AndyZe** [comment]
I think what happened was, the student opened a terminal with Ctrl+Alt+t and started typing. Then he clicked back in RViz to change the view, accidentally sending a base motion command. I've also seen people accidentally press 'm' which hides the interactive marker. Then they can't move the robot.

I think your global flag idea sounds promising. I should have some bandwidth in about a week.

**Comment by rhaschke** [comment]
> I think your global flag idea sounds promising. I should have some bandwidth in about a week.

Great. Thanks a lot. You probably want to start off my already rebased branch: https://github.com/rhaschke/rviz/tree/toggle-hotkeys

**Comment by rhaschke** [comment]
Closing in favour of #1363.

**Pull request closed without merging**